### PR TITLE
Deals with 'get_affine' function to be compatible with recent nibabel release

### PR DESCRIPTION
**Pull request by KamalakerDadi**
All is in title

### CODE REVIEW

File: doc/manipulating_images/input_output.rst [hunk 1]
```diff
@@ -216,7 +219,7 @@ objects":
 
   * A string with a file path to a Nifti or Analyse image
   * An ``SpatialImage`` from nibabel, ie an object exposing ``get_data()``
-    and ``get_affine()`` methods, typically a ``Nifti1Image`` from nibabel_.
+    and ``affine`` methods, typically a ``Nifti1Image`` from nibabel_.
```
Comment (AlexandreAbraham) [inline doc/manipulating_images/input_output.rst]: `affine` is not a method, it's an attribute.

File: nilearn/_utils/compat.py [hunk 2]
```diff
@@ -58,3 +62,33 @@ def md5_hash(string):
...
         return m.hexdigest()
+
+
+def get_affine(img):
+    """A function to return affine upon nibabel version.
+
+    nibabel > 2.0.0 returns img.affine otherwise img.get_affine()
+    """
+    if isinstance(img, _basestring):
+        img = nibabel.load(img)
```
Comment (AlexandreAbraham) [inline nilearn/_utils/compat.py]: We should not need this.

File: nilearn/_utils/compat.py [hunk 3]
```diff
@@ -58,3 +62,33 @@ def md5_hash(string):
...
+
+    if LooseVersion(nibabel.__version__) > LooseVersion('2.0.0'):
+        affine = img.affine
```
Comment (AlexandreAbraham) [inline nilearn/_utils/compat.py]: checking the version at each call is not the best practice. You should put a condition on the declaration of the function at module level:
```python
if LooseVersion(nibabel.__version__) > LooseVersion('2.0.0'):
    def get_affine(img):
        return img.affine
else:
    def get_affine(img):
        return img.get_affine()
```

Comment (KamalakerDadi) [inline nilearn/_utils/compat.py]: Meaning it shouldn't be in _utils/compat.py ?

Comment (KamalakerDadi) [inline nilearn/_utils/compat.py]: Should I address in every function wherever necessary ?

**Comment by KamalakerDadi** [comment]
Any comments about the hack in the file ``` nilearn/plotting/displays.py ``` adjusting for MNI template and normal Nifti image ?

File: nilearn/_utils/compat.py (hunk 3 above)
Comment (KamalakerDadi) [inline nilearn/_utils/compat.py]: Ok, then I am going to address this in every module.

Comment (AlexandreAbraham) [inline nilearn/_utils/compat.py]: This should stay in this module. Look at how python 2 / python 3 versions are handled:

```python
if sys.version_info[0] == 3:
    # Python 3 declarations
else:
   # Python 2 declarations
```

You should do the same here:
```python
if LooseVersion(nibabel.__version__) > LooseVersion('2.0.0'):
    # Code compatible with newest versions
else:
    # Code compatible with older versions
```

File: nilearn/plotting/displays.py [hunk 4]
```diff
@@ -640,9 +640,12 @@ def _map_show(self, img, type='imshow',
...
                 data = np.ma.masked_inside(data, -threshold, threshold,
                                            copy=False)
-            img = new_img_like(img, data, img.get_affine())
+            img = new_img_like(img, data, _utils.compat.get_affine(img))
 
-        affine = img.get_affine()
+        if img.__class__.__name__ == '_MNI152Template':
+            affine = img.get_affine()
+        else:
+            affine = _utils.compat.get_affine(img)
```
Comment (AlexandreAbraham) [inline nilearn/plotting/displays.py]: You don't need this: both `img.get_affine()` and `img.affine` will work with _MNI152Template.

File: nilearn/_utils/compat.py (hunk 3 above)
Comment (KamalakerDadi) [inline nilearn/_utils/compat.py]: Ok. I will change it.

File: nilearn/plotting/displays.py (hunk 4 above)
Comment (KamalakerDadi) [inline nilearn/plotting/displays.py]: I don't think so. _MNI152Template has a function ```get_affine``` in plotting module. It returns empty when I tried to call with this PR function. Hence tests are failing completely.

File: nilearn/_utils/niimg_conversions.py [hunk 5]
```diff
@@ -297,7 +297,7 @@ def check_niimg_3d(niimg, dtype=None):
         See http://nilearn.github.io/manipulating_images/input_output.html.
         If niimg is a string, consider it as a path to Nifti image and
         call nibabel.load on it. If it is an object, check if get_data()
-        and get_affine() methods are present, raise TypeError otherwise.
+        and affine attributes are present, raise TypeError otherwise.
```
Comment (bthirion) [inline nilearn/_utils/niimg_conversions.py]: check if the get_data() method and affine attribute are present

**Comment by bthirion** [comment]
LGTM apart for small documentation glitch.

File: nilearn/plotting/displays.py (hunk 4 above)
Comment (KamalakerDadi) [inline nilearn/plotting/displays.py]: I might be wrong also. Let's wait for any other opinion.

Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: The code that you have above is a bit hacky. A better way to solve this issue would be to fix the "affine" attribute of the _MNI152Template. The way you can do this is using Python "properties", they are functions that are called upon access of an attribute. It is actually done in the _MNI152Template for the 'shape' attribute, so you have to mimic the strategy used for "shape".

Comment (KamalakerDadi) [inline nilearn/plotting/displays.py]: Sure. I will look into it.

File: nilearn/_utils/compat.py [hunk 6]
```diff
@@ -58,3 +62,19 @@ def md5_hash(string):
...
+    def get_affine(img):
+        return img.get_affine()
+
+
+if LooseVersion(nibabel.__version__) >= LooseVersion('2.0.0'):
+    def get_header(img):
+        return img.header
+else:
+    def get_header(img):
+        return img.get_header()
```
Comment (GaelVaroquaux) [inline nilearn/_utils/compat.py]: Complete nitpick (sorry, I miss it earlier): could you please join the 2 if/else blocks above into one, that defines the 2 functions at once.

**Comment by GaelVaroquaux** [comment]
Aside from my nitpick, this looks good to me. Thanks!

**Comment by GaelVaroquaux** [comment]
LGTM. Merging. Thanks!

**Pull request merged**